src/goto-programs/goto_model.h

```diff
@@ -67,4 +67,61 @@ class goto_modelt
   void unload(const irep_idt &name) { goto_functions.unload(name); }
 };
 
+/// Interface providing access to a single function in a GOTO model, plus its
+/// associated symbol table.
+/// Its purpose is to permit GOTO program renumbering (a non-const operation on
+/// goto_functionst) without providing a non-const reference to the entire
+/// function map. For example, incremental function loading uses this, as in
+/// that situation functions other than the one currently being loaded should
+/// not be altered.
+class goto_model_functiont
+{
+public:
+  /// Construct a location numbering interface
+  /// \param goto_model: will be used to ensure unique numbering of
+  ///   goto programs, specifically incrementing its unused_location_number
+  ///   member each time a program is re-numbered.
+  explicit goto_model_functiont(
+    goto_modelt &goto_model, goto_functionst::goto_functiont &goto_function):
+  goto_model(goto_model),
+  goto_function(goto_function)
+  {
+  }
+
+  /// Re-number our goto_function. After this method returns all instructions'
+  /// location numbers may have changed, but will be globally unique and in
+  /// program order within the program.
```

Comment:
Nit pick: "program order" may be ambiguous (or possibly wrong) here; the numbers are strictly increasing as they are laid out in the underlying storage (source text or a list of statements), which need not coincide with the order of execution.

Reply:
This is the wording used elsewhere in `./goto-programs/goto_program_template.h` so I'll keep this as-is